### PR TITLE
Add test utils for reentrant redis locks

### PR DESCRIPTION
**Pull request by millerdev**
Some tests were waiting for lock timeouts due to double acquire. The nose plugin will cause this situation to error out, and those failures should be able to be resolved with `reentrant_redis_locks` decorator/context manager.

`./manage.py test corehq.apps.data_interfaces.tests.test_auto_case_updates`
before: Ran 35 tests in 294.054s
after: Ran 35 tests in 50.866s

The offending test in that module was `CaseRuleOnSaveTests.test_run_on_save`, which does two reentrant locks.

### CODE REVIEW

File: corehq/tests/locks.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    Usage as context manager:
+
+        def test_something():
+            with reentrant_redis_locks():
+                ...
+    """
+    def get_reentrant_lock(key, **kw):
+        try:
+            lock = locks[key]
```
Comment (czue) [inline corehq/tests/locks.py:34]: i found it a little confusing that `locks` was defined below this. is that a normal pattern for scope closures?

File: corehq/tests/locks.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    manager = context()
+    return manager if test is None else manager(test)
+
+
+@attr.s
+class TestRedisClient:
+    lock = attr.ib()
+
+    def __call__(self):
+        return self
```
Comment (czue) [inline corehq/tests/locks.py:58]: for my own education, what's going on here?

File: corehq/tests/noseplugins/redislocks.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        assert _LOCK.acquire(blocking=False), \
+            f"{uniform_description(case.test)}: concurrent tests not supported"
+        self.patch.start()
+
+    def stopTest(self, case):
+        self.patch.stop()
+        _LOCK.release()
+
+
+class TestRedisClient:
```
Comment (czue) [inline corehq/tests/noseplugins/redislocks.py:37]: what's the difference between this class and `TestRedisClient` in `locks.py`? assume it's by design that they have the same name? If not, may be worth renaming one of them for clarity

File: corehq/tests/test_locks.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def test_redislocks_nose_plugin():
+    lock1 = get_redis_lock(__name__, timeout=0.5, name="test")
+    assert isinstance(lock1.lock, TestLock), lock1.lock
+    lock1.acquire()
+    lock2 = get_redis_lock(__name__, timeout=0.5, name="test")
+    with assert_raises(TimeoutError):
+        lock2.acquire()
+    # lock1.release() -> LockError: Cannot release a lock that's no longer owned
```
Comment (czue) [inline corehq/tests/test_locks.py:18]: is this line meant to be a form of documentation of some kind? if the intention is that `lock1.release()` should throw a `LockError` could that not be part of the test?

**Review by czue — APPROVED** [review]
This code looks great, although admittedly I don't fully have my head wrapped around it. Most of my questions / comments are to better understand what's going on.

Build times are :fire: though.

File: corehq/tests/locks.py (hunk 1 above)
Comment (millerdev) [reply corehq/tests/locks.py:34]: I think I adopted this pattern from working on JavaScript where it's a best practice to define functions above where they are called since JS hoists them anyway (and there are weird edge cases in JS where bad things happen if you don't do that).

In any case, I don't think this is a bad pattern in Python, and there are advantages to grouping together statements that are executed one after the other (and keep function definitions either before or after as scoping allows).

File: corehq/tests/locks.py (hunk 2 above)
Comment (millerdev) [reply corehq/tests/locks.py:58]: This class (`TestRedisClient`) is instantiated on line 48 in this module. When the patched version of `get_redis_client` is called it invokes this `__call__` method, which must return an object that has a `lock()` method.

File: corehq/tests/noseplugins/redislocks.py (hunk 3 above)
Comment (millerdev) [reply corehq/tests/noseplugins/redislocks.py:37]: Thanks for pointing that out. I can likely unify them so they use the same class. I'll do that in a follow-up PR.

File: corehq/tests/locks.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+log = logging.getLogger(__name__)
+
+
+def reentrant_redis_locks(test=None):
+    """Decorator/context manager to enable reentrant redis locks
+
+    This is useful for tests that do things like acquire a lock and then
+    fire off a celery task (which will usually be executed synchronously
+    in tests) before the lock is released.
```
Comment (dannyroberts) [inline corehq/tests/locks.py:18]: Scanning the code it looks like this basically loosens the behavior to say it's okay for a lock to be acquired multiple times at once, as long as _by the end_ when the context manager exists, no one's still holding the lock.

That also jives with the wikipedia definition of reentrant locks https://en.wikipedia.org/wiki/Reentrant_mutex (which yes I did have to look up) in case that helps anyone else as well.

Daniel, anything to add to how I described it, or is that pretty much right?

**Review by dannyroberts — APPROVED** [review]
Awesome find, and looks like it puts the offending build job _way_ under the travis limit now. @millerdev we owe you a debt of gratitude!

File: corehq/tests/test_locks.py (hunk 4 above)
Comment (millerdev) [reply corehq/tests/test_locks.py:18]: Yes, it's documentation to show why the lock, which is acquired above, is not released in this test (because it gets implicitly released on timeout). I'll change it to use `assert_raises` in a follow-up PR.

File: corehq/tests/locks.py (hunk 5 above)
Comment (millerdev) [reply corehq/tests/locks.py:18]: Yes,  what you said is correct. I will update this docstring a bit since it's not entirely clear as is (blame it on writing docstrings hastily at 10:00pm). I'll improve it in a follow-up PR.

**Comment by millerdev** [comment]
I'm still not convinced that this totally fixes the travis timeout issues we've been seeing lately. The one test that was dramatically improved as a result of this has been that way for a _long_ time, and I don't think its runtime changed at all recently.

File: corehq/tests/locks.py (hunk 5 above)
Comment (millerdev) [reply corehq/tests/locks.py:18]: https://github.com/dimagi/commcare-hq/pull/27402/commits/3f2cc9c05c43544bc66a5e2d358bc802c2e3fe02

File: corehq/tests/test_locks.py (hunk 4 above)
Comment (millerdev) [reply corehq/tests/test_locks.py:18]: https://github.com/dimagi/commcare-hq/pull/27402/commits/5e95fc2f85cfdf7602b8c8a442262d4b1f5285fc

File: corehq/tests/noseplugins/redislocks.py (hunk 3 above)
Comment (millerdev) [reply corehq/tests/noseplugins/redislocks.py:37]: https://github.com/dimagi/commcare-hq/pull/27402/commits/2f5497487a13066bbd95f0aa091c0c240986777d
